src/app/utils/MGPCanFail.ts

```diff
@@ -0,0 +1,76 @@
+import { Comparable, comparableEquals } from './Comparable';
+import { MGPOptional } from './MGPOptional';
+
+export abstract class MGPCanFail<T extends Comparable> {
+    public static success<T extends Comparable>(value: NonNullable<T>): MGPCanFail<T> {
+        if (value == null) throw new Error('CanFail cannot be created with empty value, use MGPOptional.empty instead');
+        return new MGPCanFailSuccess(value);
+    }
+    public static failure<T extends Comparable>(reason: string): MGPCanFail<T> {
+        return new MGPCanFailFailure(reason);
```

Comment:
reason can be null ?

Reply:
C'est mieux pas effectivement, maintenant c'est NonNullablé

Reply:
NonNullable ça n'empêche donc pas de passer null/undefined à la compil ou au lint, mais ça throw ?
C'est ça ? (et, est-ce vérifié ?)

Reply:
En théorie, NonNullable ça empêche que tu fasses failure(null), ou failure(variable) sans être sûr que variable soit pas null.
Pas de throw; ça compile pas si tu fais du caca. Cependant il semble que ça accepte undefined! D'où l'option "strictNullCheck" qui devrait être préférée, où tout est considéré comme non nullable et non undefinable, sauf quand tu l'autorise (sain par défaut donc, plutôt que le comportement inverse)

Reply:
Ah oui, ça sera bien ça!